Add nginx image

jmaupetit opened the pull request:
- [x] Add OpenShift-compatible `nginx` image Dockerfile
- [x] Add `build` and `publish` scripts
- [x] Add documentation
- [x] Automate the builds with CircleCI (upon tagging, _e.g._ `nginx-1.13`)

jmaupetit committed:
🐳(nginx) add Dockerfile for nginx service

Each service Dockerfile should inherit from an official (and maintained)
image, and add new instructions that make this new image compatible with
OpenShift.

The expected path for a new service is:
docker/images/<SERVICE>/Dockerfile

All files eventually required during the build should stand in:
docker/images/<SERVICE>/

jmaupetit committed:
✨(scripts) add build & publish scripts

Add bash scripts to automate service image builds and publication.

* `build`: usage: `bin/build SERVICE`

Check that a Dockerfile exists for this service (it is expected to be in
`docker/images/<SERVICE>/Dockerfile`), get the service tag and build an
image tagged with our dockerhub namespace (fundocker), an image prefix
(openshift), the service name and the original image tag, _e.g._

fundocker/openshift-nginx:1.13

* `publish`: usage: `bin/publish SERVICE`

If the current release of the service image has been build, push it to
DockerHub.

jmaupetit committed:
📝(readme) add documentation to work with images

* Introduce bin/ scripts usage and philosophy
* Add instruction to add a new service image
* Add a list of maintained images/services

cchaudier commented on README.md:
```diff
@@ -0,0 +1,114 @@
+### Publish a service image
+
+Once built, you can publish your image to [DockerHub](https://hub.docker.com)
+via the `bin/publish` script, _e.g._:
```
The CI don't publish on docker hub ?

jmaupetit replied:
Not when you reviewed this work, but now it does :wink:

jmaupetit committed:
👷(circle) add build and publication jobs

We always build all services to ensure nothing is broken in a PR. A
service is published when the git repository has been tagged with the
following pattern: <service>-<version> (e.g. nginx-1.13).

To publish all services, we need to tag the git repository with the
following pattern: all-<date> (e.g. all-20180420).

cchaudier commented on README.md:
````diff
@@ -0,0 +1,114 @@
+$ mkdir docker/images/foo
+```
+
+Then write OpenShift compatibility statements that will add new Docker layers
````
A little link for the OpenShift compatibility requirements ? ;)

cchaudier commented on bin/_utils.sh:
```diff
@@ -0,0 +1,90 @@
+    local service=$1
+    local service_image_path="${BASE_SERVICE_IMAGE_PATH}/${service}/Dockerfile"
+
+    echo "$service_image_path"
```
why not ? 
`echo "${BASE_SERVICE_IMAGE_PATH}/${service}/Dockerfile"`

cchaudier commented on bin/_utils.sh:
```diff
@@ -0,0 +1,90 @@
+    if [[ -z $1 ]]; then
+        echo "check_service_image_path: service name is missing"
+        exit 1
+    fi
```
Dry way : 
You can add this function : 
```
_check_service_name() {
  if [[ -z $2 ]]; then
    echo "$1: service name is missing"
    exit 1
  fi
}
```

And call it like this : 
`_check_service_name _check_service_image_path $*`

cchaudier commented on bin/build:
```diff
@@ -0,0 +1,69 @@
+    local service=$1
+    local image_path=$(_get_service_image_path $service)
+    local version=$(_get_base_image_tag $service)
+    local tag=$(_get_target_image_tag $service)
```
tag is use on all script why don't init a global var in _utils.sh ?

jmaupetit replied:
I prefer resolving it dynamically to avoid possible collisions.

cchaudier commented on bin/_utils.sh:
```diff
@@ -0,0 +1,90 @@
+# Get target image tag (fully qualified, e.g. namespace/name:tag)
+#
+# Usage: _get_target_image_tag SERVICE
+function _get_target_image_tag() {
```
is not target_image_full_name ?

cchaudier commented on bin/_utils.sh:
```diff
@@ -0,0 +1,90 @@
+        exit 1
+    fi
+    local service=$1
+    local version=$(_get_base_image_tag $service)
```
Version is docker tag ?
IMHO it can create misunderstanding.

jmaupetit replied:
Yes it's a docker tag.

jmaupetit replied:
I've changed to name to avoid any confusion.

cchaudier commented on bin/publish:
```diff
@@ -0,0 +1,64 @@
+    echo
+    echo "Arguments:"
+    echo "  SERVICE: the name of the service to publish (e.g. nginx)"
+}
```
it's the same usage of the build script no ?
it can be _usage function in _utils.sh ?

jmaupetit replied:
It would be over-engineered IMO. Let's keep scripts separated and use our lib for core functions. Deal?

cchaudier replied:
If you want or you can do this for the default usage : 
```
function usage() {
  _usage
}
```
And override usage when it's not the default ;-)

I know it's overkill :$

jmaupetit replied:
Yup. I prefer having self sufficient scripts: `usage` and `main` functions should explicitly be part of it. It's less DRY, but it's way more simpler to understand. Deal?

cchaudier reviewed (commented):
Hello, greate job.
I've some questions and I propose some improvements.

jmaupetit committed:
👌(scripts) be more explicit and DRY

Address @cchaudier's comments. Hat tip to him!

cchaudier commented on bin/_utils.sh:
```diff
@@ -7,30 +7,31 @@ declare IMAGE_NAME_PREFIX="openshift-"
+    fi
+}
+
+
```
`alias _check_service_name='_check_service_name ${FUNCNAME} $*'` # add alias for DRY way

cchaudier commented on bin/_utils.sh:
```diff
@@ -7,30 +7,31 @@ declare IMAGE_NAME_PREFIX="openshift-"
-        echo "get_service_image_path: service name is missing"
-        exit 1
-    fi
+    _check_service_argument ${FUNCNAME[0]} $*
```
`_check_service_argument`

cchaudier commented on bin/_utils.sh:
```diff
@@ -0,0 +1,85 @@
+#!/usr/bin/env bash
+
+set -eo pipefail
+
```
`shopt -s expand_aliases` # enable alias expansion

cchaudier commented on bin/_utils.sh:
```diff
@@ -0,0 +1,85 @@
+#
+# Usage: _get_service_image_path SERVICE
+function _get_service_image_path() {
+    _check_service_argument ${FUNCNAME[0]} $*
```
`    _check_service_argument` #minimum call

cchaudier reviewed (commented):
Some new comments and improvement
